### ISSUE TITLE
Allow for the downloading of reels

### ISSUE THREAD

**Issue by jathayde** [instaloader/instaloader#812]
**Provide us a use case of the feature**
Currently a user can download various types of content from an Instagram account, such as IGTV, Stories, Highlights, and Photos. IN early August, 2020, Instagram added "Reels" to compete with TikTok. As a user can currently download any of the other content types, it is desirable to be able to completely archive an account, including reels, from one command. Per the goal listed in the contributing guidelines ("to mimic the browser's behavior to access the data that is available through the Instagram web interface") this seems like the only missing functionality.

**Describe the solution you'd like**
`instaloader --reels <accountname>`, working in conjunction with other switches, would download to a folder entitled "Reels" (or just inline as MP4s with the rest of the content, per the switch dictating output string format). If a user has no reels, it would return a "No reels are available for this user." message and continue through the rest of the download chain. 

**Describe alternatives you've considered**
There are various tools that will do this one at a time, but nothing with the efficiency of Instaloader from the command line. 

**If the feature request is accepted, would you be willing to submit a pull request?**
No development experience with Python, but have done a good amount of SW dev (UXer who knows Ruby/Rails and some JS), and happy to collaborate as I can.

**Additional context**
N/A

**Comment by Thammus** [comment]
Thanks for the feature suggestion. It would really be nice to be able to download reels with Instaloader. Since the web interface of Instagram does not yet provide access to reels, it would be necessary to use mobile endpoints in order to implement this feature.

**Comment by ungenio** [comment]
> 
> 
> Thanks for the feature suggestion. It would really be nice to be able to download reels with Instaloader. Since the web interface of Instagram does not yet provide access to reels, it would be necessary to use mobile endpoints in order to implement this feature.

It is now possible to view Reels from the web interface!
Please implement this feature.

**Comment by jimcat** [comment]
The path for reels is the same as for posts (i.e. https://www.instagram.com/p/XXXXXXXXXXX/ -- where XXXXXXXXXXX is the post/reel ID), hence with the current version reels can be downloaded individually with the same command as if they were an individual post: 

`instaloader -- -XXXXXXXXXXX `

The only inconvenient is that you must know the reel ID by manually checking it from the website -- i.e. cannot automate download all reels for a list of users, etc.

**Comment by Poloin34** [comment]
Use [@Poloin.downloader](https://instagram.com/poloin.downloaderl) to download reels :p

**Comment by nzec** [comment]
This is the request to get the reels:
```
curl 'https://i.instagram.com/api/v1/clips/user/' \
  -H 'x-ig-app-id: 567067343352427' \
  -H 'user-agent: Instagram 121.0.0.29.119 Android (26/8.0.0; 480dpi; 1080x1920; samsung; SM-G930F; herolte; samsungexynos8890; en_US; 185203708)' \
  -H 'x-csrftoken: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' \
  --data-raw 'target_user_id=1301203788&page_size=12' \
```
(app-id and user-agent copied from [here](https://github.com/dilame/instagram-private-api/blob/623a348343e34058c3a286693740aa3698aed3cc/src/core/state.ts#L120))
(add your own csrftoken)
I didn't find `x-ig-app-id` in this repo but it seems to be necessary for this query to work.

How should I go about implementing this?

**Comment by Thammus** [comment]
Nice finding! Instaloader is not using Android requests so far. However, this curl invocation could be translated to work with requests. The X-CSRFToken header is already set in the session that Instaloader uses for requests when logged in.

**Comment by nzec** [comment]
```py
headers = { 'user-agent': 'Instagram 146.0.0.27.125 (iPhone12,1; iOS 13_3; en_US; en-US; scale=2.00; 1656x3584; 190542906)' }
data = {
  'target_user_id': '1902228491',
  'page_size': '12'
}
response = requests.post('https://i.instagram.com/api/v1/clips/user/', headers=headers, data=data)
```

I think the best would be to make a new function in `instaloadercontext.py` that performs a post request (or modify the `get_iphone_json` (the existing one doesn't work because it is necessary to make a post request and pass in the data as `data=` and not `params=`)) and make a function `get_reels` in the class `Instaloader` in `instaloader.py`

**Comment by Thammus** [comment]
This looks promising. Besides the new function in `instaloadercontext.py` which issues the post request it would be helpful to have a method `get_reels()` inside the `Profile` class to iterate over all available reels of the given profile. So the next step after successfully requesting the data would be to add functionality into `structures.py` and maybe parse this reels data into a format that can be represented as an `instaloader.Post`.

**Comment by mdPlusPlus** [comment]
> Since the web interface of Instagram does not yet provide access to reels

As far as I can tell the web interface now provides access to reels.

**Comment by tari3x** [comment]
Isn't it true that every reel is also a regular post? I see that both links exists:
```
https://www.instagram.com/reel/$ID
https://www.instagram.com/p/$ID
```

And they both point to the same video.

**Comment by Fak3** [comment]
[gallery-dl](https://github.com/mikf/gallery-dl) can download reels: 
`gallery-dl --cookies-from-browser firefox "https://www.instagram.com/%username%/reels/"`

**Comment by uliar** [comment]
Still no reels support? That's most of the recent Instagram content.

**Comment by tari3x** [comment]
See my comment above - every real is also a regular post, you don't need anything special for it.

**Comment by tari3x** [comment]
Yes, they are the same, therefore there's no need to support reels specially in instaloader. You can just download all the posts.

**Comment by premium9000** [comment]
> Yes, they are the same, therefore there's no need to support reels specially in instaloader. You can just download all the posts.

dude, it's not the same. if the instagram user doesn't add the reels to their posts feed, then it stays in the reel feed only. When you run instaloader on the user's account, it doesn't pull the reels into the download. 

Instaloader still only downloads posts and highlights, but not any videos from the REELS page.

**Comment by tari3x** [comment]
Can you point to an example of a reel that is not a post?

**Comment by premium9000** [comment]
> Can you point to an example of a reel that is not a post?

https://www.instagram.com/avakittv/reels/ these reels are not in the posts. so when you use instaloader, it'll download the posts, but not the reels.

**Comment by jpfleischer** [comment]
here is my semi solution https://github.com/instaloader/instaloader/pull/2294/files

**Comment by premium9000** [comment]
> here is my semi solution #2294

amazing, thank you. I never checked to see if you replied. Before reading your response, I learned how to install and use instaloader and that worked to download reels.

**Comment by jpfleischer** [comment]
> > here is my semi solution #2294
> 
> amazing, thank you. I never checked to see if you replied. Before reading your response, I learned how to install and use instaloader and that worked to download reels.

@premium9000 , maybe you are mistaken? this _is_ instaloader, and you meant to write something else? in any case, please do share how you achieved downloading reels

**Comment by premium9000** [comment]
> > > here is my semi solution #2294
> > 
> > 
> > amazing, thank you. I never checked to see if you replied. Before reading your response, I learned how to install and use instaloader and that worked to download reels.
> 
> @premium9000 , maybe you are mistaken? this _is_ instaloader, and you meant to write something else? in any case, please do share how you achieved downloading reels

oh I'm sorry. so not sure if this is allowed, but I downloaded TamperMonkey and although this isn't a mass auto-downloading option, I installed IG Helper (https://github.com/SN-Koarashi/ig-helper) and I'm downloading reels through google chrome with it.
